sdk/identity/identity/src/credentials/managedIdentityCredential/msalMsiProvider.ts

```diff
@@ -0,0 +1,291 @@
+          const isAvailable = await imdsMsi.isAvailable({
+            scopes,
+            clientId: this.clientId,
+            getTokenOptions: options,
+            identityClient: this.isAvailableIdentityClient,
+            resourceId: this.resourceId,
+          });
+
+          if (!isAvailable) {
+            throw new CredentialUnavailableError(
+              `ManagedIdentityCredential: The managed identity endpoint is not available.`,
+            );
+          }
+        }
+
+        // If we got this far, it means:
+        // - This is not a tokenExchangeMsi,
+        // - We already probed for IMDS endpoint availability and failed-fast if it's unreachable.
+        // We can proceed normally by calling MSAL for a token.
+        logger.getToken.info("Calling into MSAL for managed identity token.");
+        const token = await this.managedIdentityApp.acquireToken({
+          resource,
+        });
+
+        this.ensureValidMsalToken(scopes, token, options);
+        logger.getToken.info(formatSuccess(scopes));
+
+        return {
+          expiresOnTimestamp: token.expiresOn.getTime(),
+          token: token.accessToken,
+        };
+      } catch (err: any) {
+        logger.getToken.error(formatError(scopes, err));
+
+        // AuthenticationRequiredError described as Error to enforce authentication after trying to retrieve a token silently.
+        // TODO: why would this _ever_ happen considering we're not trying the silent request in this flow?
+        if (err.name === "AuthenticationRequiredError") {
+          throw err;
+        }
+
+        if (isNetworkError(err)) {
+          throw new CredentialUnavailableError(
+            `ManagedIdentityCredential: Network unreachable. Message: ${err.message}`,
+          );
+        }
+
+        throw new CredentialUnavailableError(
+          `ManagedIdentityCredential: Authentication failed. Message ${err.message}`,
+        );
+      }
+    });
+  }
+
+  /**
+   * Ensures the validity of the MSAL token
+   */
+  private ensureValidMsalToken(
+    scopes: string | string[],
+    msalToken?: MsalToken,
+    getTokenOptions?: GetTokenOptions,
+  ): asserts msalToken is ValidMsalToken {
+    const createError = (message: string): Error => {
+      logger.getToken.info(message);
+      return new AuthenticationRequiredError({
+        scopes: Array.isArray(scopes) ? scopes : [scopes],
+        getTokenOptions,
+        message,
+      });
+    };
+    if (!msalToken) {
+      throw createError("No response");
+    }
+    if (!msalToken.expiresOn) {
+      throw createError(`Response had no "expiresOn" property.`);
+    }
+    if (!msalToken.accessToken) {
+      throw createError(`Response had no "accessToken" property.`);
+    }
+  }
+}
+
+function isNetworkError(err: any): boolean {
+  if (err.errorCode === "network_error") {
+    return true;
+  }
+
+  // TODO: validate this is still needed or delete
```

Comment:
I think we'll still need to keep this check. Since Docker behavior will still be same? But we may have to add it under isAvailable for imds too if we don't have it already.
Adding @christothes to verify.

Reply:
This would occur for the probe request - are we checking this there, or for the 2nd IMDS attempt?

Reply:
Right now this handler is shared between the direct probe request and MSAL request. I think what makes sense here is to separate them out - handle the probe request errors separately. I will create an issue to do that before GA